library/src/c_api/c_file.rs

```diff
@@ -0,0 +1,228 @@
+use std::io::{Read, Seek};
+
+use crate::{ExternalFileProvider, ReadSeek};
+
+use super::FileCallbacks;
+
+struct CFile {
+    file_callbacks: FileCallbacks,
+    handle: *mut libc::c_void,
+}
+
+#[derive(Clone, Debug)]
+pub struct CFileProvder {
+    pub file_callbacks: FileCallbacks,
+}
+
+impl ExternalFileProvider for CFileProvder {
+    fn open(&self) -> anyhow::Result<Box<dyn ReadSeek>> {
+        let handle = (self.file_callbacks.open)();
+        let file = CFile {
+            file_callbacks: self.file_callbacks,
+            handle,
+        };
+        Ok(Box::new(file))
+    }
+}
+
+impl ReadSeek for CFile {}
+
+impl Drop for CFile {
+    fn drop(&mut self) {
+        (self.file_callbacks.close)(self.handle);
+    }
+}
+
+impl Read for CFile {
+    fn read(&mut self, buf: &mut [u8]) -> std::io::Result<usize> {
+        Ok((self.file_callbacks.read)(
+            self.handle,
+            buf.as_mut_ptr(),
+            buf.len(),
+        ))
+    }
+}
+
+impl Seek for CFile {
+    fn seek(&mut self, pos: std::io::SeekFrom) -> std::io::Result<u64> {
+        let (offset, whence) = match pos {
+            std::io::SeekFrom::Start(offset) => (offset as i64, libc::SEEK_SET),
+            std::io::SeekFrom::End(offset) => (offset, libc::SEEK_END),
+            std::io::SeekFrom::Current(offset) => (offset, libc::SEEK_CUR),
+        };
+        let result = (self.file_callbacks.seek)(self.handle, offset, whence);
+        if result < 0 {
+            Err(std::io::Error::new(
+                std::io::ErrorKind::Other,
+                format!("CFile seek failed with error code: {}", result),
+            ))
+        } else {
+            Ok(result as u64)
+        }
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use serial_test::serial;
+
+    use super::*;
+
+    static mut OPEN_CALL_COUNT: usize = 0;
```

Comment:
I guess I'm confused why these are static, instead of held off the handle pointer?  Doesn't matter a ton, but this will cause you to have to run tests in serial rather than letting them run parallel as they do by default.

Reply:
The `FileCallbacks` functions are static, and I'm not sure there's a way to associate a `self` with them. Happy to be wrong on this one

Reply:
Yeah, it would need to be held off the handle.  You're correct, it's just test code and it's fine as is.